src/System.Private.ServiceModel/tools/test/SelfHostWcfService/TestResources/DuplexResources.cs

```diff
@@ -57,4 +57,38 @@ protected override Binding GetBinding()
             return new NetTcpBinding(SecurityMode.None) { PortSharingEnabled = false };
         }
     }
+
+    internal class DuplexCallbackDataContractComplexTypeResource : EndpointResource<WcfDuplexService, IWcfDuplexService_DataContract>
+    {
+        protected override string Protocol { get { return BaseAddressResource.Tcp; } }
+
+        protected override int GetPort(ResourceRequestContext context)
+        {
+            return context.BridgeConfiguration.BridgeTcpPort + 4;
+        }
+
+        protected override string Address { get { return "tcp-nosecurity-callback"; } }
+
+        protected override Binding GetBinding()
+        {
+            return new NetTcpBinding(SecurityMode.None) { PortSharingEnabled = false };
+        }
+    }
+
+    internal class DuplexCallbackXmlComplexTypeResource : EndpointResource<WcfDuplexService, IWcfDuplexService_Xml>
+    {
+        protected override string Protocol { get { return BaseAddressResource.Tcp; } }
+
+        protected override int GetPort(ResourceRequestContext context)
+        {
+            return context.BridgeConfiguration.BridgeTcpPort + 5;
+        }
```

Comment:
I see you found my new way of choosing TCP ports :smile: .  Not now, but let's consider making some enum somewhere that gives them names.